File: Informatica/4°ANNO/C++/Le Basi/Frazione/Frazione.cpp
#include <iostream>
#include <math.h>
using namespace std;

class Frazione
{
private:
    int numeratore;
    int denominatore;

public:
    Frazione(){};

    void setNumeratore(int a)
    {
        numeratore = a;
    }

    void setDenominatore(int a)
    {
        denominatore = a;
    }

    int getNumeratore()
    {
        return numeratore;
    }

    int getDenominatore()
    {
        return denominatore;
    }

    void setNumDen(int a, int b)
    {
        numeratore = a;
        denominatore = b;
    }

    int semplifica()
    {
        int r;

        r = trovaMCD(numeratore, denominatore);

        return r;
    }

    int trovaMCD(int x, int y)
    {
        int r;

        if (x < y)
        {
            int z = x;
            x = y;
            y = z;
        }

        do
        {
            r = x % y;

            if (r != 0)
            {
                x = y;
                y = r;
            }

        } while (r != 0);

        return y;
    }

    void stampa()
    {
        float x = (float)numeratore, y = (float)denominatore;

        cout << "\n"
             << numeratore << "\n";

        if (numeratore > denominatore)
        {
            for (int i = 0; numeratore % (int)pow(10, i) != numeratore; i++)
            {
                cout << "-";
            }
        }
        else
        {
            for (int i = 0; denominatore % (int)pow(10, i) != denominatore; i++)
            {
                cout << "-";
            }
        }

        cout << " = " << x / y << "\n"
             << denominatore;
    }
};

int main()
{
    Frazione f1, f2;

    int n1, n2;

    cout << "Inserire numeratore frazione 1: ";
    cin >> n1;
    cout << "Inserire denominatore frazione 1: ";
    cin >> n2;
    f1.setNumeratore(n1);
    f1.setDenominatore(n2);
    cout << "Inserire numeratore frazione 2: ";
    cin >> n1;
    cout << "Inserire denominatore frazione 2: ";
    cin >> n2;
    f2.setNumeratore(n1);
    f2.setDenominatore(n2);

    f1.stampa();
    cout << "\n";
    f2.stampa();

    cout << "\n\nL'MCD della prima frazione e': " << f1.semplifica();
    cout << "\nL'MCD della seconda frazione e': " << f2.semplifica() << endl;

    return 0;
}

File: TPSIT/4°anno/Java/ES_03/src/Operazioni/Frazione.java
package Operazioni;

public class Frazione {
    // Campi
    private int numeratore;
    private int denominatore;
    // Costruttore

    public Frazione(int numeratore, int denominatore) {
        this.numeratore = numeratore;
        this.denominatore = denominatore;
    }

    // Metodo per ottenere il numeratore
    public int getNumeratore() {
        return numeratore;
    }

    // Metodo per ottenere il denominatore
    public int getDenominatore() {
        return denominatore;
    }

    // Metodo per ottenere la rappresentazione testuale della frazione
    @Override
    public String toString() {
        return numeratore + "/" + denominatore;
    }

    // Metodo per semplificare la frazione 
    private Frazione semplifica(Frazione fraction)
    {
        int div;
        
        do
        {
            div=calcolaMCD(fraction.numeratore, fraction.denominatore);
            fraction.numeratore/=div;
            fraction.denominatore/=div;
        }while(div!=1);

        return fraction;
    }
    // Metodo per trovare l'MCD fra 2 numeri
    private int calcolaMCD(int x, int y) {

        int r;

        if (x < y) {
            int z = x;
            x = y;
            y = z;
        }

        do {
            r = x % y;

            if (r != 0) {
                x = y;
                y = r;
            }

        } while (r != 0);

        return y;
    }

    // Metodo per sommare due frazioni
    public Frazione somma(Frazione altraFrazione) {
        int nuovoNumeratore = this.numeratore * altraFrazione.denominatore
                + altraFrazione.numeratore * this.denominatore;
        int nuovoDenominatore = this.denominatore * altraFrazione.denominatore;

        Frazione nuovaFrazione = semplifica(new Frazione(nuovoNumeratore, nuovoDenominatore));

        return nuovaFrazione;
    }
    // Metodo per sottrarre due frazioni
    public Frazione sottrai(Frazione altraFrazione) {
        int nuovoNumeratore = this.numeratore * altraFrazione.denominatore
                - altraFrazione.numeratore * this.denominatore;
        int nuovoDenominatore = this.denominatore * altraFrazione.denominatore;

        Frazione nuovaFrazione = semplifica(new Frazione(nuovoNumeratore, nuovoDenominatore));

        return nuovaFrazione;
    }
    // Metodo per moltiplicare due frazioni
    public Frazione moltiplica(Frazione altraFrazione) {
        int nuovoNumeratore = this.numeratore * altraFrazione.numeratore;
        int nuovoDenominatore = this.denominatore * altraFrazione.denominatore;

        Frazione nuovaFrazione = semplifica(new Frazione(nuovoNumeratore, nuovoDenominatore));

        return nuovaFrazione;
    }
    // Metodo per dividere due frazioni
    public Frazione dividi(Frazione altraFrazione) {
        int nuovoNumeratore = this.numeratore * altraFrazione.denominatore;
        int nuovoDenominatore = this.denominatore * altraFrazione.numeratore;

        Frazione nuovaFrazione = semplifica(new Frazione(nuovoNumeratore, nuovoDenominatore));

        return nuovaFrazione;
    }
}
